### DISCUSSION TITLE
`imessage-exporter` Performance

### DISCUSSION THREAD

**Discussion by musa11971** [ReagentX/imessage-exporter#247] [Q&A] [answered]
Hi, thanks for creating this tool. I've been trying to use it to export my chats, but keep running into significant performance issues. The amount of time it would take me to export makes it simply not feasible to finish, see the following:

```
⠉ [21m] [>-------------------] 9999/429340 (0.1784/s, ETA: 4w)
```

or a bit later...

```
⠁ [23m] [>-------------------] 10395/429340 (0.0198/s, ETA: 35w)                     
```

Do note the huge chat size of 33 GB. I did not expect the export to be fast at all, but this seems extreme.
(`Estimated export size: 33.68 GB`)

Running on MacBook Pro M1 Max (32 GB) 2021, full disk access enabled.

Any ideas on what I could do to speed this up? It's a bummer I can't do it in batches. E.g. run the export for a bit today, and then continue tomorrow etc. I found that if I close my Mac and come back to it later, the export process will freeze completely and not make any progress.

**Comment by ReagentX** [comment, marked as answer]
First, check the performance notes [here](https://github.com/ReagentX/imessage-exporter/blob/develop/docs/faq.md#how-fast-is-imessage-exporter).

This doesn't make a lot of sense to me, it takes about 12 seconds on my machine to export more than 500k messages from my personal db.  On my M1 Max MacBook Pro, `imessage-exporter` reaches  36k messages per second, nowhere near your `0.0198/s`.

Did you recently delete a large conversation? I noticed that while iCloud syncs that deletion, the table gets very slow to query.

If you are converting attachments, especially large videos, the time estimate may become inflated.

Can you provide details on your machine, (cpu, os, etc) as well as the `imessage-exporter` command you are using?

**Reply by e40** [reply]
I just did an export, with the very newest version (iMessage Exporter 2.2.2).  It took about 20 minutes.  At one point I could see all my cores (Mac Studio Max M1) were going full tilt and the rate went to 0.  4y's the estimate, at one point.

Came here to ask: what's the resource bottleneck here?  @ReagentX

**Reply by ReagentX** [reply]
`imessage-exporter` is not multi-threaded, so it can only ever saturate a single core. However, the converter tools it uses to convert attachments may use all available resources. For example, transcoding large videos with `ffmpeg` will use a lot of CPU and probably slow down the export significantly.

What `imessage-exporter` command you are using?

**Reply by e40** [reply]
Using this:

```
imessage-exporter -o "$PWD" -f html -c full --no-lazy
```

I did another export at the beginning of Feb and I had updated Homebrew and saw a new version so I wondered if it would be faster.  It was not.

Is macOS throttling something, somewhere?

**Reply by ReagentX** [reply]
As explained above, `-c full` uses `ffmpeg`, which will slow down the export. This is mentioned in the [FAQ](https://github.com/ReagentX/imessage-exporter/blob/develop/docs/faq.md#how-fast-is-imessage-exporter) as well.

**Reply by e40** [reply]
Does `-c clone` do the fastest copy?

**Reply by e40** [reply]
Answered myself.  Took a few seconds with `-c clone`.

**Reply by e40** [reply]
The output using `-c clone` looks perfectly fine.  I will use that.  Thanks for the replies!

**Reply by ReagentX** [reply]
Yes, the difference between the options is outlined in [the instructions](https://github.com/ReagentX/imessage-exporter/blob/develop/imessage-exporter/README.md#how-to-use).
